From 3caa2c047d98ff6ae69e6e3b22b0b40f6f01c748 Mon Sep 17 00:00:00 2001
From: Nishant Sharma <nishant.sharma@arm.com>
Date: Thu, 4 Apr 2024 23:28:28 +0100
Subject: [PATCH] neoverse-rd: add sds banks for the remote chips

In an multichip environment, each SCP write the pcie configuration
information in is local Secure SRAM memory. BL2 can initialise all the
banks, read the data and concatenate it.

Add the support to initialize and read the data from all the SDS banks
and program it to the non secure bootloader fdt.

Signed-off-by: Nishant Sharma <nishant.sharma@arm.com>
Change-Id: Ibacccfbee09e96920b4d99ad21468f2562d63864

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../common/include/nrd2/nrd_css_fw_def2.h     |  4 +-
 .../common/include/nrd2/nrd_plat_arm_def2.h   | 25 +++++--
 .../common/include/nrd3/nrd_css_fw_def3.h     |  5 +-
 .../common/include/nrd3/nrd_plat_arm_def3.h   | 23 ++++++-
 .../board/neoverse_rd/common/nrd_image_load.c | 67 ++++++++++++-------
 plat/arm/board/neoverse_rd/common/nrd_plat2.c | 14 +++-
 plat/arm/board/neoverse_rd/common/nrd_plat3.c | 42 +++++++++++-
 7 files changed, 142 insertions(+), 38 deletions(-)

diff --git a/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_css_fw_def2.h b/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_css_fw_def2.h
index 481632bfbd25..31888a560e3d 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_css_fw_def2.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_css_fw_def2.h
@@ -19,9 +19,9 @@
 #define NRD_CSS_BL1_RW_SIZE		UL(64 * 1024)	/* 64KB */
 
 #if TRUSTED_BOARD_BOOT
-# define NRD_CSS_BL2_SIZE		UL(0x20000)
+# define NRD_CSS_BL2_SIZE		UL(0x30000)
 #else
-# define NRD_CSS_BL2_SIZE		UL(0x14000)
+# define NRD_CSS_BL2_SIZE		UL(0x30000)
 #endif
 /*
  * Since BL31 NOBITS overlays BL2 and BL1-RW, PLAT_ARM_MAX_BL31_SIZE is
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_plat_arm_def2.h b/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_plat_arm_def2.h
index 917209cfea5c..aa0a486773cc 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_plat_arm_def2.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd2/nrd_plat_arm_def2.h
@@ -72,7 +72,7 @@
 # define PLAT_ARM_MMAP_ENTRIES		U(8)
 # define MAX_XLAT_TABLES		U(5)
 #elif defined(IMAGE_BL2)
-# define PLAT_ARM_MMAP_ENTRIES		(11 + (NRD_CHIP_COUNT - 1))
+# define PLAT_ARM_MMAP_ENTRIES		(14 + (NRD_CHIP_COUNT - 1))
 
 /*
  * MAX_XLAT_TABLES entries need to be doubled because when the address width
@@ -81,7 +81,7 @@
  * > 40 bits
  *
  */
-# define MAX_XLAT_TABLES		(7  + ((NRD_CHIP_COUNT - 1) * 2))
+# define MAX_XLAT_TABLES		(10  + ((NRD_CHIP_COUNT - 1) * 2))
 #elif !USE_ROMLIB
 # define PLAT_ARM_MMAP_ENTRIES		U(11)
 # define MAX_XLAT_TABLES		U(7)
@@ -277,14 +277,31 @@ ENABLE_FEAT_RAS && FFH_SUPPORT
 #define SDS_PCIE_MMAP_IDX			U(129)
 
 #define MAX_RP_PER_CONTROLLER			4
-#if (NRD_PLATFORM_VARIANT == 0)
+#if ((NRD_PLATFORM_VARIANT == 0) || (NRD_PLATFORM_VARIANT == 3))
 #define MAX_NUM_PCIE_CONTROLLER	        	4
 #else
 #define MAX_NUM_PCIE_CONTROLLER			1
 #endif
 
+#define NRD_CSS_SDS_BASE(n) \
+	(PLAT_ARM_SDS_MEM_BASE + NRD_REMOTE_CHIP_MEM_OFFSET(n))
+
 /* Index of SDS region used in the communication with SCP */
-#define SDS_SCP_AP_REGION_ID		U(0)
+#if !defined(__ASSEMBLER__)
+typedef enum sds_region_id {
+	SDS_SCP_AP_CHIP0_REGION_ID,
+#if NRD_CHIP_COUNT > 1
+	SDS_SCP_AP_CHIP1_REGION_ID,
+#endif
+#if NRD_CHIP_COUNT > 2
+	SDS_SCP_AP_CHIP2_REGION_ID,
+#endif
+#if NRD_CHIP_COUNT > 3
+	SDS_SCP_AP_CHIP3_REGION_ID,
+#endif
+	SDS_SCP_AP_REGION_ID_MAX,
+} sds_region_id_t;
+#endif
 
 /*******************************************************************************
  * Flash config
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
index 8a08add74b0e..051f4e6a3979 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
@@ -21,7 +21,7 @@
 #define NRD_CSS_BL1_RO_BASE		NRD_CSS_SHARED_SRAM_BASE
 #define NRD_CSS_BL1_RO_SIZE		UL(0x00019000)
 
-# define NRD_CSS_BL2_SIZE		UL(0x30000)
+# define NRD_CSS_BL2_SIZE		UL(0x40000)
 
 /*
  * Since BL31 NOBITS overlays BL2 and BL1-RW, PLAT_ARM_MAX_BL31_SIZE is
@@ -81,8 +81,9 @@
 			NRD_CSS_GPC_SMMUV3_SIZE,			\
 			MT_DEVICE | MT_RW | EL3_PAS)
 
-#define NRD_CSS_IO_VIRT_SMMUV3_MMAP					\
+#define NRD_CSS_IO_VIRT_SMMUV3_MMAP(n)					\
 		MAP_REGION_FLAT(					\
+			NRD_REMOTE_CHIP_MEM_OFFSET(n) +			\
 			NRD_CSS_IO_VIRT_SMMUV3_BASE,			\
 			NRD_CSS_IO_VIRT_SMMUV3_SIZE,			\
 			MT_DEVICE | MT_RW | EL3_PAS)
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
index b1f7a257f65b..c51d53434e5f 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
@@ -62,8 +62,8 @@
 # define PLAT_ARM_MMAP_ENTRIES		U(8)
 # define MAX_XLAT_TABLES		U(5)
 #elif defined(IMAGE_BL2)
-# define PLAT_ARM_MMAP_ENTRIES		(16 + (NRD_CHIP_COUNT - 1))
-# define MAX_XLAT_TABLES		(11  + ((NRD_CHIP_COUNT - 1) * 2))
+# define PLAT_ARM_MMAP_ENTRIES		(19 + (NRD_CHIP_COUNT - 1))
+# define MAX_XLAT_TABLES		(14  + ((NRD_CHIP_COUNT - 1) * 2))
 #else
 # define PLAT_ARM_MMAP_ENTRIES		U(7)
 # define MAX_XLAT_TABLES		U(7)
@@ -367,8 +367,25 @@
 #define MAX_NUM_PCIE_CONTROLLER			1
 #endif
 
+#define NRD_CSS_SDS_BASE(n) \
+	(PLAT_ARM_SDS_MEM_BASE + NRD_REMOTE_CHIP_MEM_OFFSET(n))
+
+#if !defined(__ASSEMBLER__)
 /* Index of SDS region used in the communication with SCP */
-#define SDS_SCP_AP_REGION_ID		U(0)
+typedef enum sds_region_id {
+	SDS_SCP_AP_CHIP0_REGION_ID,
+#if NRD_CHIP_COUNT > 1
+	SDS_SCP_AP_CHIP1_REGION_ID,
+#endif
+#if NRD_CHIP_COUNT > 2
+	SDS_SCP_AP_CHIP2_REGION_ID,
+#endif
+#if NRD_CHIP_COUNT > 3
+	SDS_SCP_AP_CHIP3_REGION_ID,
+#endif
+	SDS_SCP_AP_REGION_ID_MAX,
+} sds_region_id_t;
+#endif
 
 /*******************************************************************************
  * SMMUv3 Config
diff --git a/plat/arm/board/neoverse_rd/common/nrd_image_load.c b/plat/arm/board/neoverse_rd/common/nrd_image_load.c
index bdfe892f76b1..ff2a24dcfe5b 100644
--- a/plat/arm/board/neoverse_rd/common/nrd_image_load.c
+++ b/plat/arm/board/neoverse_rd/common/nrd_image_load.c
@@ -23,8 +23,9 @@
 
 /* Information about the ECAM, MMIOL, MMIOH carveouts and Bus start and end */
 #define MAX_PCIE_BLOCK_SIZE (sizeof(struct pcie_block_mmap_list) + \
-	(MAX_NUM_PCIE_CONTROLLER * (sizeof (struct pcie_block_mmio_mmap) + \
-		(MAX_RP_PER_CONTROLLER * sizeof(struct pcie_ep_config)))))
+	((NRD_CHIP_COUNT * MAX_NUM_PCIE_CONTROLLER) * \
+	 (sizeof (struct pcie_block_mmio_mmap) + \
+	(MAX_RP_PER_CONTROLLER * sizeof(struct pcie_ep_config)))))
 
 struct pcie_mem_carveout {
 	uint64_t start;
@@ -114,28 +115,48 @@ void plat_arm_nrd_pcie_smmu_init(struct pcie_block_mmap_list *mmap_list)
 /* Copy PCIe memmap data from sds to a local buffer. */
 void plat_arm_nrd_get_pcie_mmap(struct pcie_block_mmap_list *mmap_list)
 {
+	struct {
+		uint64_t block_count;
+		uint64_t table_size;
+	} header;
+	sds_region_id_t region_id;
 	int ret;
 
-	ret = sds_init(SDS_SCP_AP_REGION_ID);
-	if (ret != SDS_OK) {
-		ERROR("SDS initialization failed. ret: %d\n", ret);
-		panic();
-	}
+	mmap_list->block_count = 0;
+	mmap_list->table_size = sizeof(header);
+	for (region_id = SDS_SCP_AP_CHIP0_REGION_ID;
+			region_id < SDS_SCP_AP_REGION_ID_MAX;
+			region_id++ ) {
 
-	ret = sds_struct_read(SDS_SCP_AP_REGION_ID,
-			      SDS_PCIE_MMAP_IDX, 0, mmap_list,
-			      sizeof(struct pcie_block_mmap_list),
-			      SDS_ACCESS_MODE_CACHED);
-	if (ret != SDS_OK) {
-		ERROR("SDS PCIE MMAP header read failed. ret: %d\n", ret);
-	}
+		ret = sds_init(region_id);
+		if (ret != SDS_OK) {
+			ERROR("SDS initialization failed. ret: %d\n", ret);
+			panic();
+		}
 
-	ret = sds_struct_read(SDS_SCP_AP_REGION_ID,
-			      SDS_PCIE_MMAP_IDX, 0, mmap_list,
-			      mmap_list->table_size,
-			      SDS_ACCESS_MODE_CACHED);
-	if (ret != SDS_OK) {
-		ERROR("SDS PCIE MMAP header read failed. ret: %d\n", ret);
+		ret = sds_struct_read(region_id,
+				SDS_PCIE_MMAP_IDX, 0, &header,
+				sizeof(header),
+				SDS_ACCESS_MODE_CACHED);
+		if (ret != SDS_OK) {
+			ERROR("SDS PCIE MMAP header read failed. ret: %d\n", ret);
+			continue;
+		}
+		INFO("Block count: 0x%"PRIx64"\n", header.block_count);
+		INFO("Table Size: 0x%"PRIx64"\n", header.table_size);
+
+		header.table_size -= sizeof(header);
+		ret = sds_struct_read(region_id,
+				SDS_PCIE_MMAP_IDX, sizeof(header),
+				(void *)((uint8_t *)mmap_list + mmap_list->table_size),
+				header.table_size,
+				SDS_ACCESS_MODE_CACHED);
+		if (ret != SDS_OK) {
+			ERROR("SDS PCIE MMAP header read failed. ret: %d\n", ret);
+			continue;
+		}
+		mmap_list->block_count += header.block_count;
+		mmap_list->table_size += header.table_size;
 	}
 }
 
@@ -152,13 +173,13 @@ void plat_arm_nrd_get_isolated_cpu_list(struct isolated_cpu_mpid_list *list)
 {
 	int ret;
 
-	ret = sds_init(SDS_SCP_AP_REGION_ID);
+	ret = sds_init(SDS_SCP_AP_CHIP0_REGION_ID);
 	if (ret != SDS_OK) {
 		ERROR("SDS initialization failed, error: %d\n", ret);
 		panic();
 	}
 
-	ret = sds_struct_read(SDS_SCP_AP_REGION_ID,
+	ret = sds_struct_read(SDS_SCP_AP_CHIP0_REGION_ID,
 			SDS_ISOLATED_CPU_LIST_ID, 0, &list->num_entries,
 			sizeof(list->num_entries), SDS_ACCESS_MODE_CACHED);
 	if (ret != SDS_OK) {
@@ -177,7 +198,7 @@ void plat_arm_nrd_get_isolated_cpu_list(struct isolated_cpu_mpid_list *list)
 		return;
 	}
 
-	ret = sds_struct_read(SDS_SCP_AP_REGION_ID,
+	ret = sds_struct_read(SDS_SCP_AP_CHIP0_REGION_ID,
 			SDS_ISOLATED_CPU_LIST_ID,
 			sizeof(list->num_entries),
 			&list->mpid_list,
diff --git a/plat/arm/board/neoverse_rd/common/nrd_plat2.c b/plat/arm/board/neoverse_rd/common/nrd_plat2.c
index 4a7b02147cc9..8f97cae0765f 100644
--- a/plat/arm/board/neoverse_rd/common/nrd_plat2.c
+++ b/plat/arm/board/neoverse_rd/common/nrd_plat2.c
@@ -45,12 +45,15 @@ const mmap_region_t plat_arm_mmap[] = {
 	NRD_ROS_SYSTEM_PERIPH_MMAP,
 	ARM_MAP_NS_DRAM1,
 #if NRD_CHIP_COUNT > 1
+	NRD_CSS_SHARED_RAM_MMAP(1),
 	NRD_ROS_MEMCNTRL_MMAP(1),
 #endif
 #if NRD_CHIP_COUNT > 2
+	NRD_CSS_SHARED_RAM_MMAP(2),
 	NRD_ROS_MEMCNTRL_MMAP(2),
 #endif
 #if NRD_CHIP_COUNT > 3
+	NRD_CSS_SHARED_RAM_MMAP(3),
 	NRD_ROS_MEMCNTRL_MMAP(3),
 #endif
 #if ARM_BL31_IN_DRAM
@@ -180,7 +183,16 @@ void plat_arm_secure_wdt_stop(void)
 }
 
 static sds_region_desc_t nrd_sds_regions[] = {
-	{ .base = PLAT_ARM_SDS_MEM_BASE },
+	{ .base = NRD_CSS_SDS_BASE(0) },
+#if NRD_CHIP_COUNT > 1
+	{ .base = NRD_CSS_SDS_BASE(1) },
+#endif
+#if NRD_CHIP_COUNT > 2
+	{ .base = NRD_CSS_SDS_BASE(2) },
+#endif
+#if NRD_CHIP_COUNT > 3
+	{ .base = NRD_CSS_SDS_BASE(3) },
+#endif
 };
 
 sds_region_desc_t *plat_sds_get_regions(unsigned int *region_count)
diff --git a/plat/arm/board/neoverse_rd/common/nrd_plat3.c b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
index 779a9a90949b..e51686dc825e 100644
--- a/plat/arm/board/neoverse_rd/common/nrd_plat3.c
+++ b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
@@ -47,7 +47,25 @@ const mmap_region_t plat_arm_mmap[] = {
 #endif
 	NRD_CSS_GPT_L1_DRAM_MMAP,
 	NRD_CSS_RMM_REGION_MMAP,
-	NRD_CSS_IO_VIRT_SMMUV3_MMAP,
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(0),
+#if NRD_CHIP_COUNT > 1
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(1),
+#endif
+#if NRD_CHIP_COUNT > 2
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(2),
+#endif
+#if NRD_CHIP_COUNT > 3
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(3),
+#endif
+#if NRD_CHIP_COUNT > 1
+	NRD_CSS_SHARED_RAM_MMAP(1),
+#endif
+#if NRD_CHIP_COUNT > 2
+	NRD_CSS_SHARED_RAM_MMAP(2),
+#endif
+#if NRD_CHIP_COUNT > 3
+	NRD_CSS_SHARED_RAM_MMAP(3),
+#endif
 	{0}
 };
 #endif /* IMAGE_BL2 */
@@ -64,7 +82,16 @@ const mmap_region_t plat_arm_mmap[] = {
 	NRD_CSS_GPT_L1_DRAM_MMAP,
 	NRD_CSS_EL3_RMM_SHARED_MEM_MMAP,
 	NRD_CSS_GPC_SMMU_SMMUV3_MMAP,
-	NRD_CSS_IO_VIRT_SMMUV3_MMAP,
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(0),
+#if NRD_CHIP_COUNT > 1
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(1),
+#endif
+#if NRD_CHIP_COUNT > 2
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(2),
+#endif
+#if NRD_CHIP_COUNT > 3
+	NRD_CSS_IO_VIRT_SMMUV3_MMAP(3),
+#endif
 #if RESET_TO_BL31
 	NRD_CSS_MAP_BL31_DTB,
 #endif
@@ -96,7 +123,16 @@ void plat_arm_secure_wdt_stop(void)
 }
 
 static sds_region_desc_t nrd_sds_regions[] = {
-	{ .base = PLAT_ARM_SDS_MEM_BASE },
+	{ .base = NRD_CSS_SDS_BASE(0) },
+#if NRD_CHIP_COUNT > 1
+	{ .base = NRD_CSS_SDS_BASE(1) },
+#endif
+#if NRD_CHIP_COUNT > 2
+	{ .base = NRD_CSS_SDS_BASE(2) },
+#endif
+#if NRD_CHIP_COUNT > 3
+	{ .base = NRD_CSS_SDS_BASE(3) },
+#endif
 };
 
 sds_region_desc_t *plat_sds_get_regions(unsigned int *region_count)
